readthedocs/projects/tasks.py

```diff
@@ -1825,37 +1826,68 @@ def finish_inactive_builds():
     """
     Finish inactive builds.
 
-    A build is consider inactive if it's not in ``FINISHED`` state and it has been
-    "running" for more time that the allowed one (``Project.container_time_limit``
-    or ``DOCKER_LIMITS['time']`` plus a 20% of it).
+    A build is consider inactive if all the following are true:
 
-    These inactive builds will be marked as ``success`` and ``FINISHED`` with an
+    - it's not in ``FINISHED`` state
+    - it was created +15 minutes ago
+    - there is not task queued for it on redis
+    - celery is not currently executing it
+
+    These inactive builds will be marked as ``success=False`` and ``FINISHED`` with an
     ``error`` to be communicated to the user.
     """
-    # TODO similar to the celery task time limit, we can't infer this from
-    # Docker settings anymore, because Docker settings are determined on the
-    # build servers dynamically.
-    # time_limit = int(DOCKER_LIMITS['time'] * 1.2)
-    # Set time as maximum celery task time limit + 5m
-    time_limit = 7200 + 300
+    time_limit = 15 * 60  # 15 minutes
     delta = datetime.timedelta(seconds=time_limit)
     query = (
         ~Q(state=BUILD_STATE_FINISHED) & Q(date__lte=timezone.now() - delta)
     )
 
-    builds_finished = 0
+    stale_build_pks = []
     builds = Build.objects.filter(query)[:50]
+    redis_client = redis.Redis.from_url(settings.BROKER_URL)
+
     for build in builds:
+        build_stale = True
+
+        # 1. check if it's being executed by celery
+        for queue, tasks in app.control.inspect().active():
```

Comment:
I remember that I tested this and was accurate. I don't know how trustable it may be, tho.

Another quick test today in production showed accurate numbers at least:

```python
# number of builds marked as running in the db / numbers of builds executed by workers
In [43]: (Build.objects.exclude(state__in=('finished', 'triggered')).count(), sum([len(tasks) for queue, tasks in app.control.inspect().active
    ...: ().items() if queue.startswith('build')]))
Out[43]: (17, 17)
```

I don't think this will be 100% accurate at all times, tho. Mainly because the DB is immediate and the worker takes a small time to _start the queued task_. So, there may be a difference there. However, this should only happen at the edges (when the task is about to start or about to end) and both cases should be covered. The "about to start" because we have a minimum delta time running and the "about to end" because the task will still be present in Celery but not in DB --which won't make the algorithm to mark it as stale.